table/tables/mutation_checker.go

```diff
@@ -0,0 +1,244 @@
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/rowcodec"
+	"go.uber.org/zap"
+)
+
+type mutation = struct {
+	key   kv.Key
+	flags kv.KeyFlags
+	value []byte
+}
+
+type indexHelperInfo = struct {
+	indexInfo   *model.IndexInfo
+	rowColInfos []rowcodec.ColInfo
+}
+
+// CheckIndexConsistency checks whether the given set of mutations corresponding to a single row is consistent.
+// Namely, assume the database is consistent before, applying the mutations shouldn't break the consistency.
+// It aims at reducing bugs that will corrupt data, and preventing mistakes from spreading if possible.
+//
+// Assume the set of row values changes from V1 to V2. We check
+// (1) V2 - V1 = {added indices}
+// (2) V1 - V2 = {deleted indices}
+//
+// To check (1), we need
+// (a) {added indices} is a subset of {needed indices} => each index mutation is consistent with the input/row key/value
+// (b) {needed indices} is a subset of {added indices}. The check process would be exactly the same with how we generate
+// 		the mutations, thus ignored.
+func CheckIndexConsistency(sessVars *variable.SessionVars, t *TableCommon,
+	dataAdded, dataRemoved []types.Datum, memBuffer kv.MemBuffer, sh kv.StagingHandle) error {
+	sc := sessVars.StmtCtx
+	if sh == 0 {
+		return nil
+	}
+	mutations := collectTableMutationsFromBufferStage(t, memBuffer, sh)
+	if err := checkRowAdditionConsistency(sessVars, t.Meta().Columns, dataAdded, mutations); err != nil {
+		return errors.Trace(err)
+	}
+	if err := checkIndexKeys(sc, sessVars, t, dataAdded, dataRemoved, mutations); err != nil {
+		return errors.Trace(err)
+	}
+	// TODO: check whether handles match in index and row mutations
+	return nil
+}
+
+// checkIndexKeys checks whether the decoded data from keys of index mutations are consistent with the expected ones.
+func checkIndexKeys(sc *stmtctx.StatementContext, sessVars *variable.SessionVars, t *TableCommon,
+	dataAdded []types.Datum, dataRemoved []types.Datum, mutations []mutation) error {
+	indexIDMap := make(map[int64]indexHelperInfo)
+	for _, index := range t.indices {
+		indexIDMap[index.Meta().ID] = indexHelperInfo{
+			index.Meta(),
+			BuildRowcodecColInfoForIndexColumns(index.Meta(), t.Meta()),
+		}
+	}
+
+	for _, m := range mutations {
+		if !tablecodec.IsIndexKey(m.key) {
+			continue
+		}
+
+		_, indexID, _, err := tablecodec.DecodeIndexKey(m.key)
+		if err != nil {
+			continue
+		}
+
+		indexHelperInfo, ok := indexIDMap[indexID]
+		if !ok {
+			return errors.New("index not found")
+		}
+
+		// when we cannot decode the key to get the original value
+		if len(m.value) == 0 && NeedRestoredData(indexHelperInfo.indexInfo.Columns, t.Meta().Columns) {
+			continue
+		}
+
+		decodedIndexValues, err := tablecodec.DecodeIndexKV(m.key, m.value, len(indexHelperInfo.indexInfo.Columns),
+			tablecodec.HandleNotNeeded, indexHelperInfo.rowColInfos)
```

Comment:
Will index key utilities like `expression index`,  `collations` or `row formats` introduce corner cases here?
Need help /cc @lysu @wjhuang2016

Reply:
No, I think it's fine here.